src/Expander.hs

```diff
@@ -456,34 +459,36 @@ initializeKernel = do
               Just _ ->
                 error "TODO throw real error - already expanding a module"
               Nothing -> do
-                bodyPtr <- liftIO newModBodyPtr
+                bodyPtr <- liftIO newDeclTreePtr
                 modifyState $ set expanderModuleTop (Just bodyPtr)
                 Stx _ _ (_ :: Syntax, name, imports, body, exports) <- mustHaveEntries stx
                 _actualName <- mustBeIdent name
 
                 resolveImports imports
 
-                forkExpandDecls bodyPtr body
+                vp <- newDeclValidityPtr
+                expandDeclForms bodyPtr mempty vp body
 
                 buildExports exports
                 pure ()
         )
       ]
 
-    declPrims :: [(Text, Scope -> DeclPtr -> DeclValidityPtr -> Syntax -> Expand ())]
+    declPrims :: [(Text, DeclTreePtr -> ScopeSet -> DeclValidityPtr -> Syntax -> Expand ())]
```

Comment:
I took the opportunity to reorder a few type signatures to be more consistent: the destination is always first, and the syntax is always last.

Reply:
Excellent!